build/Dockerfile

```diff
@@ -29,7 +29,7 @@ ARG USER_UID
 
 RUN dnf remove -y --disableplugin=subscription-manager \
         subscription-manager \
-    && dnf install -y python${PYTHON_VERSION} procps \
+    && dnf install -y python${PYTHON_VERSION} procps g++ python3.11-devel     \
```

Comment:
@tharapalanivel does quantization pipeline also need g++ ? wondering if that comes into GPL license

Reply:
@anhuong do you recall what errors you got when you omitted the g++ and devel deps?

@Ssukriti for your question on g++, its hard to find direct references, but for stdlib++ it should be GPLv3 https://gcc.gnu.org/onlinedocs/libstdc++/manual/license.html

Reply:
The error I hit into when just installing `fms-acceleration` and trying to run `auto_gptq` was `RuntimeError: Failed to find C compiler. Please specify via CC environment variable.` with full error:
```sh
$ python -m tuning.sft_trainer --model_name_or_path "TechxGenus/Meta-Llama-3-8B-GPTQ" --training_data_path /fmaas-integration-tests-pvc/tuning/input/cc_tone_sft_format_1000_train.json --output_dir /fmaas-integration-tests-pvc/tuning/output/anhuong/llama3-8b-gptq-lora-tone-single-gpu-run-cli --num_train_epochs 10 --per_device_train_batch_size 4 --gradient_accumulation_steps 1 --learning_rate 2e-4 --response_template "\n### Response:" --dataset_text_field output --peft_method lora --r 16 --lora_alpha 16 --lora_dropout 0.1 --auto_gptq triton_v2 --fp16 --torch_dtype float16 --target_modules "q_proj", "k_proj", "v_proj", "o_proj"
/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/triton_utils/dequant.py:143: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @custom_fwd
/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/triton_utils/dequant.py:151: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @custom_bwd
/home/tuning/.local/lib/python3.11/site-packages/huggingface_hub/file_download.py:1150: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
quantize_config.json: 100%|███████████████████████████████████████████████████████████████| 265/265 [00:00<00:00, 2.38MB/s]
INFO - Ignoring unknown parameter in the quantization configuration: is_marlin_format.
INFO - `checkpoint_format` is missing from the quantization configuration and is automatically inferred to gptq
config.json: 100%|████████████████████████████████████████████████████████████████████| 1.01k/1.01k [00:00<00:00, 9.46MB/s]
model.safetensors: 100%|██████████████████████████████████████████████████████████████| 5.74G/5.74G [09:49<00:00, 9.73MB/s]
Some weights of the model checkpoint at /home/tuning/.cache/huggingface/hub/models--TechxGenus--Meta-Llama-3-8B-GPTQ/snapshots/78292ae18b170923ae5880e99291209b760899d6/model.safetensors were not used when initializing LlamaForCausalLM: {'model.layers.22.self_attn.o_proj.bias', 'model.layers.30.mlp.up_proj.bias', 'model.layers.26.self_attn.o_proj.bias', 'model.layers.4.self_attn.v_proj.bias', 'model.layers.22.mlp.up_proj.bias', 'model.layers.7.mlp.down_proj.bias', 'model.layers.10.mlp.up_proj.bias', 'model.layers.7.self_attn.o_proj.bias', 'model.layers.9.mlp.down_proj.bias', 'model.layers.11.self_attn.k_proj.bias', 'model.layers.0.mlp.gate_proj.bias', 'model.layers.18.mlp.gate_proj.bias', 'model.layers.11.self_attn.v_proj.bias', 'model.layers.15.self_attn.q_proj.bias', 'model.layers.2.self_attn.k_proj.bias', 'model.layers.6.self_attn.o_proj.bias', 'model.layers.26.self_attn.q_proj.bias', 'model.layers.27.mlp.up_proj.bias', 'model.layers.15.mlp.gate_proj.bias', 'model.layers.20.mlp.down_proj.bias', 'model.layers.6.mlp.down_proj.bias', 'model.layers.27.self_attn.k_proj.bias', 'model.layers.18.self_attn.v_proj.bias', 'model.layers.3.self_attn.q_proj.bias', 'model.layers.23.mlp.down_proj.bias', 'model.layers.22.self_attn.v_proj.bias', 'model.layers.0.self_attn.q_proj.bias', 'model.layers.19.self_attn.v_proj.bias', 'model.layers.8.self_attn.k_proj.bias', 'model.layers.11.mlp.up_proj.bias', 'model.layers.9.mlp.up_proj.bias', 'model.layers.29.mlp.down_proj.bias', 'model.layers.10.self_attn.q_proj.bias', 'model.layers.17.mlp.gate_proj.bias', 'model.layers.3.mlp.gate_proj.bias', 'model.layers.9.mlp.gate_proj.bias', 'model.layers.30.self_attn.o_proj.bias', 'model.layers.26.mlp.up_proj.bias', 'model.layers.27.mlp.gate_proj.bias', 'model.layers.30.mlp.gate_proj.bias', 'model.layers.0.self_attn.v_proj.bias', 'model.layers.20.self_attn.k_proj.bias', 'model.layers.5.mlp.gate_proj.bias', 'model.layers.21.mlp.up_proj.bias', 'model.layers.21.self_attn.v_proj.bias', 'model.layers.23.self_attn.v_proj.bias', 'model.layers.24.mlp.down_proj.bias', 'model.layers.12.mlp.down_proj.bias', 'model.layers.21.self_attn.q_proj.bias', 'model.layers.1.self_attn.v_proj.bias', 'model.layers.21.mlp.down_proj.bias', 'model.layers.30.mlp.down_proj.bias', 'model.layers.28.self_attn.k_proj.bias', 'model.layers.20.self_attn.v_proj.bias', 'model.layers.12.self_attn.o_proj.bias', 'model.layers.14.self_attn.k_proj.bias', 'model.layers.18.self_attn.o_proj.bias', 'model.layers.18.self_attn.k_proj.bias', 'model.layers.22.mlp.down_proj.bias', 'model.layers.27.mlp.down_proj.bias', 'model.layers.31.self_attn.v_proj.bias', 'model.layers.7.self_attn.v_proj.bias', 'model.layers.10.self_attn.o_proj.bias', 'model.layers.19.mlp.up_proj.bias', 'model.layers.7.mlp.gate_proj.bias', 'model.layers.20.self_attn.o_proj.bias', 'model.layers.15.mlp.down_proj.bias', 'model.layers.3.self_attn.k_proj.bias', 'model.layers.8.self_attn.o_proj.bias', 'model.layers.18.mlp.down_proj.bias', 'model.layers.25.mlp.gate_proj.bias', 'model.layers.29.self_attn.o_proj.bias', 'model.layers.17.mlp.down_proj.bias', 'model.layers.22.self_attn.k_proj.bias', 'model.layers.14.self_attn.q_proj.bias', 'model.layers.11.mlp.gate_proj.bias', 'model.layers.10.mlp.gate_proj.bias', 'model.layers.30.self_attn.q_proj.bias', 'model.layers.15.self_attn.v_proj.bias', 'model.layers.22.self_attn.q_proj.bias', 'model.layers.28.mlp.gate_proj.bias', 'model.layers.13.self_attn.v_proj.bias', 'model.layers.2.self_attn.q_proj.bias', 'model.layers.17.self_attn.q_proj.bias', 'model.layers.19.mlp.gate_proj.bias', 'model.layers.14.mlp.up_proj.bias', 'model.layers.30.self_attn.v_proj.bias', 'model.layers.0.self_attn.o_proj.bias', 'model.layers.7.self_attn.q_proj.bias', 'model.layers.13.mlp.down_proj.bias', 'model.layers.20.mlp.gate_proj.bias', 'model.layers.1.mlp.down_proj.bias', 'model.layers.31.mlp.gate_proj.bias', 'model.layers.16.self_attn.v_proj.bias', 'model.layers.24.mlp.up_proj.bias', 'model.layers.16.mlp.down_proj.bias', 'model.layers.28.mlp.down_proj.bias', 'model.layers.7.self_attn.k_proj.bias', 'model.layers.14.self_attn.v_proj.bias', 'model.layers.17.self_attn.k_proj.bias', 'model.layers.25.self_attn.q_proj.bias', 'model.layers.25.self_attn.v_proj.bias', 'model.layers.6.self_attn.k_proj.bias', 'model.layers.29.self_attn.k_proj.bias', 'model.layers.11.mlp.down_proj.bias', 'model.layers.14.mlp.gate_proj.bias', 'model.layers.16.self_attn.k_proj.bias', 'model.layers.21.self_attn.o_proj.bias', 'model.layers.23.mlp.gate_proj.bias', 'model.layers.12.mlp.gate_proj.bias', 'model.layers.1.self_attn.k_proj.bias', 'model.layers.13.mlp.gate_proj.bias', 'model.layers.25.self_attn.o_proj.bias', 'model.layers.31.self_attn.k_proj.bias', 'model.layers.23.self_attn.k_proj.bias', 'model.layers.5.self_attn.k_proj.bias', 'model.layers.8.mlp.gate_proj.bias', 'model.layers.15.self_attn.k_proj.bias', 'model.layers.17.self_attn.v_proj.bias', 'model.layers.18.self_attn.q_proj.bias', 'model.layers.15.self_attn.o_proj.bias', 'model.layers.21.mlp.gate_proj.bias', 'model.layers.2.mlp.gate_proj.bias', 'model.layers.3.mlp.up_proj.bias', 'model.layers.24.self_attn.o_proj.bias', 'model.layers.27.self_attn.v_proj.bias', 'model.layers.28.self_attn.v_proj.bias', 'model.layers.25.mlp.up_proj.bias', 'model.layers.29.self_attn.q_proj.bias', 'model.layers.13.self_attn.o_proj.bias', 'model.layers.2.self_attn.o_proj.bias', 'model.layers.20.mlp.up_proj.bias', 'model.layers.0.mlp.up_proj.bias', 'model.layers.12.self_attn.q_proj.bias', 'model.layers.3.self_attn.v_proj.bias', 'model.layers.25.self_attn.k_proj.bias', 'model.layers.27.self_attn.o_proj.bias', 'model.layers.19.self_attn.k_proj.bias', 'model.layers.1.mlp.up_proj.bias', 'model.layers.0.mlp.down_proj.bias', 'model.layers.5.self_attn.o_proj.bias', 'model.layers.16.mlp.gate_proj.bias', 'model.layers.23.self_attn.o_proj.bias', 'model.layers.28.self_attn.q_proj.bias', 'model.layers.4.mlp.down_proj.bias', 'model.layers.24.self_attn.v_proj.bias', 'model.layers.11.self_attn.q_proj.bias', 'model.layers.9.self_attn.q_proj.bias', 'model.layers.13.self_attn.q_proj.bias', 'model.layers.6.mlp.gate_proj.bias', 'model.layers.23.mlp.up_proj.bias', 'model.layers.6.self_attn.q_proj.bias', 'model.layers.28.mlp.up_proj.bias', 'model.layers.6.mlp.up_proj.bias', 'model.layers.12.mlp.up_proj.bias', 'model.layers.29.self_attn.v_proj.bias', 'model.layers.26.mlp.gate_proj.bias', 'model.layers.1.self_attn.o_proj.bias', 'model.layers.5.self_attn.v_proj.bias', 'model.layers.26.self_attn.v_proj.bias', 'model.layers.6.self_attn.v_proj.bias', 'model.layers.10.mlp.down_proj.bias', 'model.layers.10.self_attn.k_proj.bias', 'model.layers.15.mlp.up_proj.bias', 'model.layers.16.self_attn.q_proj.bias', 'model.layers.17.self_attn.o_proj.bias', 'model.layers.18.mlp.up_proj.bias', 'model.layers.19.mlp.down_proj.bias', 'model.layers.24.self_attn.q_proj.bias', 'model.layers.24.mlp.gate_proj.bias', 'model.layers.17.mlp.up_proj.bias', 'model.layers.25.mlp.down_proj.bias', 'model.layers.0.self_attn.k_proj.bias', 'model.layers.23.self_attn.q_proj.bias', 'model.layers.2.self_attn.v_proj.bias', 'model.layers.3.mlp.down_proj.bias', 'model.layers.8.mlp.down_proj.bias', 'model.layers.20.self_attn.q_proj.bias', 'model.layers.1.self_attn.q_proj.bias', 'model.layers.5.mlp.down_proj.bias', 'model.layers.12.self_attn.v_proj.bias', 'model.layers.2.mlp.up_proj.bias', 'model.layers.8.mlp.up_proj.bias', 'model.layers.5.self_attn.q_proj.bias', 'model.layers.7.mlp.up_proj.bias', 'model.layers.19.self_attn.o_proj.bias', 'model.layers.4.mlp.gate_proj.bias', 'model.layers.9.self_attn.k_proj.bias', 'model.layers.29.mlp.gate_proj.bias', 'model.layers.19.self_attn.q_proj.bias', 'model.layers.14.self_attn.o_proj.bias', 'model.layers.9.self_attn.o_proj.bias', 'model.layers.9.self_attn.v_proj.bias', 'model.layers.4.self_attn.k_proj.bias', 'model.layers.27.self_attn.q_proj.bias', 'model.layers.30.self_attn.k_proj.bias', 'model.layers.26.mlp.down_proj.bias', 'model.layers.26.self_attn.k_proj.bias', 'model.layers.31.self_attn.q_proj.bias', 'model.layers.28.self_attn.o_proj.bias', 'model.layers.12.self_attn.k_proj.bias', 'model.layers.8.self_attn.q_proj.bias', 'model.layers.8.self_attn.v_proj.bias', 'model.layers.1.mlp.gate_proj.bias', 'model.layers.14.mlp.down_proj.bias', 'model.layers.29.mlp.up_proj.bias', 'model.layers.24.self_attn.k_proj.bias', 'model.layers.31.self_attn.o_proj.bias', 'model.layers.22.mlp.gate_proj.bias', 'model.layers.5.mlp.up_proj.bias', 'model.layers.16.mlp.up_proj.bias', 'model.layers.10.self_attn.v_proj.bias', 'model.layers.4.mlp.up_proj.bias', 'model.layers.31.mlp.down_proj.bias', 'model.layers.31.mlp.up_proj.bias', 'model.layers.3.self_attn.o_proj.bias', 'model.layers.4.self_attn.q_proj.bias', 'model.layers.13.self_attn.k_proj.bias', 'model.layers.11.self_attn.o_proj.bias', 'model.layers.13.mlp.up_proj.bias', 'model.layers.2.mlp.down_proj.bias', 'model.layers.21.self_attn.k_proj.bias', 'model.layers.16.self_attn.o_proj.bias', 'model.layers.4.self_attn.o_proj.bias'}. This may or may not be an issue - make sure that the checkpoint does not have unnecessary parameters, or that the model definition correctly corresponds to the checkpoint.
INFO - Compatibility: converting `checkpoint_format` from `gptq` to `gptq_v2`.
tokenizer_config.json: 100%|██████████████████████████████████████████████████████████| 50.9k/50.9k [00:00<00:00, 8.44MB/s]
tokenizer.json: 100%|█████████████████████████████████████████████████████████████████| 9.08M/9.08M [00:01<00:00, 6.91MB/s]
special_tokens_map.json: 100%|████████████████████████████████████████████████████████████| 301/301 [00:00<00:00, 2.86MB/s]
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
PAD token set to default, missing in tokenizer
UNK token set to default, missing in tokenizer
Generating train split: 1000 examples [00:00, 12080.19 examples/s]
Map: 100%|███████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 39309.32 examples/s]
/home/tuning/.local/lib/python3.11/site-packages/transformers/training_args.py:1847: FutureWarning: `--push_to_hub_token` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--hub_token` instead.
  warnings.warn(
/home/tuning/.local/lib/python3.11/site-packages/huggingface_hub/utils/_deprecation.py:100: FutureWarning: Deprecated argument(s) used in '__init__': dataset_text_field, max_seq_length. Will not be supported from version '1.0.0'.

Deprecated positional argument(s) used in SFTTrainer, please use the SFTConfig to set these arguments instead.
  warnings.warn(message, FutureWarning)
/home/tuning/.local/lib/python3.11/site-packages/trl/trainer/sft_trainer.py:280: UserWarning: You passed a `max_seq_length` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
  warnings.warn(
/home/tuning/.local/lib/python3.11/site-packages/trl/trainer/sft_trainer.py:318: UserWarning: You passed a `dataset_text_field` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
  warnings.warn(
Map: 100%|███████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 37903.99 examples/s]
/home/tuning/.local/lib/python3.11/site-packages/accelerate/accelerator.py:451: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
/home/tuning/.local/lib/python3.11/site-packages/accelerate/accelerator.py:488: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.scaler = torch.cuda.amp.GradScaler(**kwargs)
Detected kernel version 4.18.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.
  0%|                                                                                             | 0/2500 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/home/tuning/.local/lib/python3.11/site-packages/tuning/sft_trainer.py", line 523, in main
    train(
  File "/home/tuning/.local/lib/python3.11/site-packages/tuning/sft_trainer.py", line 337, in train
    trainer.train()
  File "/home/tuning/.local/lib/python3.11/site-packages/trl/trainer/sft_trainer.py", line 451, in train
    output = super().train(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/trainer.py", line 1780, in train
    return inner_training_loop(
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/trainer.py", line 2118, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/trainer.py", line 3036, in training_step
    loss = self.compute_loss(model, inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/trainer.py", line 3059, in compute_loss
    outputs = model(**inputs)
              ^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/accelerate/utils/operations.py", line 819, in forward
    return model_forward(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/accelerate/utils/operations.py", line 807, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/amp/autocast_mode.py", line 43, in decorate_autocast
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/peft/peft_model.py", line 1577, in forward
    return self.base_model(
           ^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/peft/tuners/tuners_utils.py", line 188, in forward
    return self.model.forward(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 1196, in forward
    outputs = self.model(
              ^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 1016, in forward
    layer_outputs = decoder_layer(
                    ^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/accelerate/hooks.py", line 169, in new_forward
    output = module._old_forward(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 739, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
                                                          ^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/transformers/models/llama/modeling_llama.py", line 435, in forward
    query_states = self.q_proj(hidden_states)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/qlinear/qlinear_tritonv2.py", line 171, in forward
    out = quant_linear_fn.apply(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/autograd/function.py", line 574, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/torch/amp/autocast_mode.py", line 455, in decorate_fwd
    return fwd(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/triton_utils/dequant.py", line 145, in forward
    output = quant_matmul_248(input, qweight, scales, qzeros, g_idx, bits, maxq)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/triton_utils/dequant.py", line 135, in quant_matmul_248
    W = dequant248(qweight, scales, qzeros, g_idx, bits, maxq=maxq)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/fms_acceleration_peft/gptqmodel/nn_modules/triton_utils/dequant.py", line 117, in dequant248
    dequant_kernel_248[grid](
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/jit.py", line 345, in <lambda>
    return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/autotuner.py", line 156, in run
    timings = {config: self._bench(*args, config=config, **kwargs) for config in pruned_configs}
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/autotuner.py", line 156, in <dictcomp>
    timings = {config: self._bench(*args, config=config, **kwargs) for config in pruned_configs}
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/autotuner.py", line 133, in _bench
    return do_bench(kernel_call, warmup=self.num_warmups, rep=self.num_reps, quantiles=(0.5, 0.2, 0.8))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/testing.py", line 103, in do_bench
    fn()
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/autotuner.py", line 114, in kernel_call
    self.fn.run(
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/jit.py", line 607, in run
    device = driver.active.get_current_device()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/driver.py", line 23, in __getattr__
    self._initialize_obj()
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/driver.py", line 20, in _initialize_obj
    self._obj = self._init_fn()
                ^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/driver.py", line 9, in _create_driver
    return actives[0]()
           ^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/backends/nvidia/driver.py", line 371, in __init__
    self.utils = CudaUtils()  # TODO: make static
                 ^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/backends/nvidia/driver.py", line 80, in __init__
    mod = compile_module_from_src(Path(os.path.join(dirname, "driver.c")).read_text(), "cuda_utils")
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/backends/nvidia/driver.py", line 57, in compile_module_from_src
    so = _build(name, src_path, tmpdir, library_dirs(), include_dir, libraries)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/build.py", line 32, in _build
    raise RuntimeError("Failed to find C compiler. Please specify via CC environment variable.")
RuntimeError: Failed to find C compiler. Please specify via CC environment variable.
```

Reply:
Note that g+ is GPL licensed but we can leverage the exceptions in the license and distribute as long as we are compiling code using standard practices and we arent modifying g++
https://www.gnu.org/licenses/gcc-exception-3.1

Reply:
If `{PYTHON_VERSION}` is not 3.11, does installing `python3.11-devel` still work?

Reply:
Good question!

Reply:
I think this should be switched to simply `python-devel` or make it consistent with `python${PYTHON_VERSION}` and do `python${PYTHON_VERSION}-devel`

Reply:
python-devel technically installs v3.9.18-3.el9_4.5 whereas python3.11-devel installed v3.11.7-1.el9_4.5 and python3.12-devel installs 3.12.1-4.el9_4.3. And setting `python${PYTHON_VERSION}-devel` only works 3.11, 3.12, and 3.9 but there isn't a python3.10-devel
```
No match for argument: python3.10-devel
Error: Unable to find a match: python3.10-devel
```

Thus I could add an if statement in here but it looks kinda bad
```
RUN dnf remove -y --disableplugin=subscription-manager \
        subscription-manager \
    && dnf install -y python${PYTHON_VERSION} procps g++ \
    && if [[ "${PYTHON_VERSION}" == "3.11" ]] || [[ "${PYTHON_VERSION}" == "3.12" ]]; \
    then dnf install -y python${PYTHON_VERSION}-devel; \
    else dnf install -y python3-devel; \
    fi \
    && ln -s /usr/bin/python${PYTHON_VERSION} /bin/python \
    && python -m ensurepip --upgrade \
    && python -m pip install --upgrade pip \
    && python -m pip install --upgrade setuptools \
    && dnf update -y \
    && dnf clean all
```